### ISSUE TITLE
Broken links running `make linkcheck`

### ISSUE THREAD

**Issue by rocky** [ethereum/solidity#6458]
I just ran `make linkcheck` in ethereum/solidity/docs and I see these broken links:
*  `(line  159) broken   ` https://github.com/ethereum/wiki/wiki/JavaScript-API#contract-events `- Anchor 'contract-events' not found`
* `(line   22) broken  ` https://semver.org/#spec-item-4 `- Anchor 'spec-item-4' not found`
* `(line  214) broken  `  https://www.visualstudio.com/downloads/#build-tools-for-visual-studio-2017 ` - Anchor 'build-tools-for-visual-studio-2017' not found`
* `(line  200) broken  `https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/FAQ.md#how-do-i-uninstall-homebrew `- Anchor 'how-do-i-uninstall-homebrew' not found`
* `line   92) broken `   https://live.ether.camp/ `- HTTPSConnectionPool(host='live.ether.camp', port=443): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x7fb13487cd30>: Failed to establish a new connection: [Errno 111] Connection refused'))`
* (line  383) broken    https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/ECRecovery.sol - 404 Client Error: Not Found for url: https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/ECRecovery.sol

Perhaps `make lintcheck` is something that might get added to an infrequent CI test?

**Comment by axic** [comment]
There are a bunch more failures now as well as redirects.

**Comment by cameel** [comment]
I did a thorough link scrubbing in #9587 and fixed most of them. The remaining ones are the visual studio links (will be fixed in #9476) and one I missed (#9707 fixes it).

There are still four more that `linkcheck` considers broken but these are just false positives.

**Comment by cameel** [comment]
> Perhaps `make lintcheck` is something that might get added to an infrequent CI test?

I wanted to add a check like that to our `b_doc` task in CI but unfortunately it's not feasible. There are too many false positives (see the end of this comment if you want the details). We would have to submit a patch to fix `linkcheck` in these cases and also refrain from using placeholders in example links. I'm not sure it's worth the effort/inconvenience. What do you think @chriseth? 

Anyway, the core of this issue will be resolved once #9707 is merged so I'm going to close it then. If we really want the CI check despite the downsides, I think we should create a separate issue specifically for it.

### Details of false positives
1) `linkcheck` does not support anchors created using the `href` attribute (reports them as broken). It requires the `id` attribute. For example this is how headings look on github:
    ``` html
    <a id="user-content-example-usage-with-smtsolver-callback" class="anchor" aria-hidden="true" href="#example-usage-with-smtsolver-callback">
    ```
    https://github.com/ethereum/solc-js#example-usage-with-smtsolver-callback (which you get when you just copy the link) is treated as broken.
    https://github.com/ethereum/solc-js#user-content-example-usage-with-smtsolver-callback is fine but you have to dig in page source to get it.

    And while for github switching to using `id` would be a workaround, https://semver.org/#spec-item-4 is more tricky. Here the same anchor is used in both `href` and `id` but `linkcheck` still sees it as broken.

    In the end, I think that the only way not to have to deal with it would be to submit a patch for the `linkcheck` module in Sphinx.
2) We have some links that are not really supposed to be valid but there's no way to disable checks for them. E.g.
    ```rst
    .. note::
        This method will result in a prerelease build leading to e.g. a flag
        being set in each bytecode produced by such a compiler.
        If you want to re-build a released Solidity compiler, then
        please use the source tarball on the github release page:

        https://github.com/ethereum/solidity/releases/download/v0.X.Y/solidity_0.X.Y.tar.gz

        (not the "Source code" provided by github).
    ```

    Actually, there's only one such place right now so it's not too bad but there could be more in the future and being able to create such links with placeholders is useful sometimes.

**Comment by axic** [comment]
I think running `make linkcheck` should be part of the (pre-)release process where we fix obvious issues.

**Comment by cameel** [comment]
@axic Yeah, this does seem like a better idea than an unforgiving automatic check.

**Comment by cameel** [comment]
Just a small clarification about the anchors. Turns out it's not really a bug or missing feature in Sphinx. It's just that these pages handle their anchors dynamically in JavaScript. Fixing it would require either running the JS during linkcheck or hard-coding the known rules on a per-site basis. Either is bad so looks like dead link detection is not something that we'll be able to fully automate in the forseeable future.

Specifically:
1. Github: Turns out that the `<a href="#xyz">` used for headings is not a standard feature and only works due to JS magic. There's already an issue about it in Sphinx: https://github.com/sphinx-doc/sphinx/issues/6779.
2. semver.org: Sphinx sees the links as broken because the HTML returned by the server has no `id`s for `<li>` elements. They're added on the client side using JS.
